File: leetcode/queue_stack/2345_Finding_the_Number_of_Visible_Mountains/codes/stack.cpp
#include <vector>
#include <stack>
#include <algorithm>
using namespace std;

class Solution {
public:
    int visibleMountains(vector<vector<int>>& peaks) {
        sort(peaks.begin(), peaks.end());

        stack<int> st;
        int end = INT_MIN;

        for (const auto& peak : peaks) {
            int left = peak[0] - peak[1];
            int right = peak[0] + peak[1];

            while (!st.empty() && st.top() >= left) {
                st.pop();
            }

            if (right > end) {
                st.push(left);
                end = right;
            }
        }

        return st.size();
    }
};

File: leetcode/queue_stack/2345_Finding_the_Number_of_Visible_Mountains/codes/stack.py
from typing import List

class Solution:
    def visibleMountains(self, peaks: List[List[int]]) -> int:
        peaks = sorted(peaks, key = lambda peak: peak[0])

        stack = []
        end = float("-inf")

        for peak in peaks:
            left = peak[0] - peak[1]
            right = peak[0] + peak[1]

            while stack and stack[-1] >= left:
                stack.pop()

            if right > end:
                stack.append(left)
                end = right

        return len(stack)
        